code/controllers/CronTaskController.php

```diff
@@ -97,16 +117,43 @@ public function index(SS_HTTPRequest $request) {
 	 * @param CronTask $task
 	 */
 	public function runTask(CronTask $task) {
-		$cron = Cron\CronExpression::factory($task->getSchedule());
-		$isDue = $this->isTaskDue($task, $cron);
-		// Update status of this task prior to execution in case of interruption
-		CronTaskStatus::update_status(get_class($task), $isDue);
-		if($isDue) {
-			$this->output(get_class($task).' will start now.');
-			$task->process();
-		} else {
-			$this->output(get_class($task).' will run at '.$cron->getNextRunDate()->format('Y-m-d H:i:s').'.');
+
+		$canRunTask = true;
+		$enforceSchedule = true;
+		$isDue = true;
+
+		if(method_exists($task, "canRunTask")) {
+			$canRunTask = $task->canRunTask();
+		}
+
+		if(method_exists($task, "enforceSchedule")) {
+			$enforceSchedule = $task->enforceSchedule();
 		}
+
+		if($canRunTask) {
+
+			if($enforceSchedule) {
+				$cron = Cron\CronExpression::factory($task->getSchedule());
+				$isDue = $this->isTaskDue($task, $cron);
+			}			
+
+			// Update status of this task prior to execution in case of interruption
+			CronTaskStatus::update_status(get_class($task), $isDue);
+
+			if($isDue) {
+				$this->output(get_class($task).' will start now.');
+				$task->process();
+			} else {
+				$this->output(get_class($task).' will run at '.$cron->getNextRunDate()->format('Y-m-d H:i:s').'.');
+			}
+
+		}
+		else {
```

Comment:
Would prefer to avoid deep nesting whenever possible :-) Would be great if you could rewrite this conditional into a guard clause - i.e. interrupt the execution if the failure mode occurs.